File: _wpf/wpf-business-programmer/4-wpf-business-programmer-m4-exercise-files/Samples/03-WPFShapesUserControl-CS/WPFShapesUserControl-CS/MidpointValueConverter.cs
using System;
using System.Globalization;
using System.Windows.Data;

namespace WPFShapesUserControl_CS
{
  public class MidpointValueConverter : IMultiValueConverter
  {
    #region Convert/ConvertBack Methods
    public object Convert(object[] values, Type targetType,
      object parameter, CultureInfo culture)
    {
      double offset = 0;
      double totalMeasure = 0;
      double controlMeasure = 0;

      // Make sure enough "values" were passed
      if (values == null || values.Length < 2)
      {
        throw new ArgumentException("The MidpointValueConverter class requires 2 double values to be passed to it. First pass the total measure (width/height) of the outside control, then the total measure (width/height) of the inside control.", "values");
      }

      // Get the total width/height
      totalMeasure = (double)values[0];
      // Get the width/height of the control
      controlMeasure = (double)values[1];

      // check for any offset to the width/height
      if (parameter != null)
        offset = System.Convert.ToDouble(parameter);

      return (object)(((totalMeasure - controlMeasure) / 2) + offset);
    }

    public object[] ConvertBack(object value, Type[] targetTypes,
      object parameter, CultureInfo culture)
    {
      throw new NotImplementedException();
    }
    #endregion
  }
}

File: _wpf/wpf-business-programmer/4-wpf-business-programmer-m4-exercise-files/Samples-2012/03-WPFShapesUserControl-VB/WPFShapesUserControl-VB/MidpointValueConverter.vb
Imports System
Imports System.Globalization
Imports System.Windows.Data

Public Class MidpointValueConverter
  Implements IMultiValueConverter

#Region "Convert/ConvertBack Methods"
  Public Function Convert(ByVal values As Object(), ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IMultiValueConverter.Convert
    Dim offset As Double = 0
    Dim totalMeasure As Double = 0
    Dim controlMeasure As Double = 0

    ' Make sure enough "values" were passed
    If values Is Nothing OrElse values.Length < 2 Then
      Throw New ArgumentException("The MidpointValueConverter class requires 2 double values to be passed to it. First pass the total measure (width/height) of the outside control, then the total measure (width/height) of the inside control.", "values")
    End If

    ' Get the total width/height
    totalMeasure = System.Convert.ToDouble(values(0))
    ' Get the width/height of the control
    controlMeasure = System.Convert.ToDouble(values(1))

    If parameter IsNot Nothing Then
      offset = System.Convert.ToDouble(parameter)
    End If

    Return DirectCast((((totalMeasure - controlMeasure) / 2) + offset), Object)
  End Function

  Public Function ConvertBack(ByVal value As Object, ByVal targetTypes As Type(), ByVal parameter As Object, ByVal culture As CultureInfo) As Object() Implements IMultiValueConverter.ConvertBack
    Throw New NotImplementedException()
  End Function
#End Region
End Class